### PR TITLE
Lab01 - Fibonacci series

### LINKED ISSUE #591: Lab1 - Fibonacci series
- Implement Fibonacci series
- Implement Negative Fibonacci series
- Add unit tests

### PR DESCRIPTION
**Pull request by harirakr**
Fixes #591 

Review of colleague's PR #587

#### Changes proposed in this PR:
- Implement Lab1 - Fibonacci and NegativeFibonacci series
- Add unit tests

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,8 @@
 # Output of the go coverage tool, specifically when used with LiteIDE
 *.out
 coverage.txt
+
+# VIM swap files
+*~
+*.swp
+*.swo
```
Comment (juliaogris) [inline .gitignore:18]: please put this in you global `~/.gitignore` (editor specific setup doesn't belong with project).

File: 01_fib/harirakr/main_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"bytes"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNegativeFib(t *testing.T) {
+
```
Comment (juliaogris) [inline 01_fib/harirakr/main_test.go]: Unnecessary blank line.
No need for blank lines at beginning or end of block.
Use blank lines between type, method and function definitions and within a block for logical separation.

Belowaswell.

File: 01_fib/harirakr/main_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	assert.Equal(t, []int{0, 1, -1, 2, -3, 5, -8, 13, -21, 34}, fib(-10))
+	assert.Equal(t, []int{0, 1, -1}, fib(-3))
+	assert.Equal(t, []int{0, 1}, fib(-2))
+	assert.Equal(t, []int{0}, fib(-1))
```
Comment (juliaogris) [inline 01_fib/harirakr/main_test.go]: Add `fib(0)`

File: 01_fib/harirakr/main_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		t.Errorf("main fail; expected=%v, actual=%v", expected, actual)
+	}
+}
+
+// Test input out of bounds
+func TestFibOutOfBounds(t *testing.T) {
+	var buf1 bytes.Buffer
+	out = &buf1
+
+	fib(93)
```
Comment (juliaogris) [inline 01_fib/harirakr/main_test.go:44]: you need to use int64 for fib calculations if you test bounds explicitly. int defaults to int32 on go playground for instance.

File: 01_fib/harirakr/main.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+var out io.Writer = os.Stdout
+
+func fib(n int) []int {
+	if n >= 93 || n <= -93 {
+		fmt.Fprintln(out, "fib: Range out of bounds; takes input > -93 and < 93")
+		return nil
+	}
+	series := make([]int, 0)
```
Comment (juliaogris) [inline 01_fib/harirakr/main.go]: `make([]int, 0, n )` you know the capacity.

File: 01_fib/harirakr/main.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	b := 0
+	for i := 0; i < n*sign; i++ {
+		sum := a + b*sign
+		if sum == 0 {
+			b = 1
+			series = append(series, a)
+			continue
+		}
+		a, b = b, sum
+		series = append(series, a)
```
Comment (juliaogris) [inline 01_fib/harirakr/main.go:27]: Initialise as `a, b := 0, 1`
Than you can collapse the loop body to
```go
a, b = b, a+sign*b
series = append(series, a)
```
This is aligned with lab requirements which (writely or not) asks you not to print the first `0` (see [lab assignment](https://github.com/anz-bank/go-course/tree/master/01_fib#lab-1---fibonacci))

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please read through the [Collected Code Review Comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments)
- [Fix your pr description](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#pr-description)
- [Limit the number of commits to one or two](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#limit-the-number-of-commits-to-one-or-two)

More comments inline.

File: 01_fib/harirakr/main_test.go (hunk 3 above)
Comment (harirakr) [reply 01_fib/harirakr/main_test.go]: Don.

File: 01_fib/harirakr/main_test.go (hunk 2 above)
Comment (harirakr) [reply 01_fib/harirakr/main_test.go]: Don.

File: 01_fib/harirakr/main.go (hunk 6 above)
Comment (harirakr) [reply 01_fib/harirakr/main.go:27]: Don.

File: 01_fib/harirakr/main.go (hunk 5 above)
Comment (harirakr) [reply 01_fib/harirakr/main.go]: Don.

File: 01_fib/harirakr/main_test.go (hunk 4 above)
Comment (harirakr) [reply 01_fib/harirakr/main_test.go:44]: Don.

File: .gitignore (hunk 1 above)
Comment (harirakr) [reply .gitignore:18]: @juliaogris While trying to correct this PR with git rebase, I think I have made a mess of this PR. I am reverting someone's changes as well. Now I am not sure how to fix this PR. Shall I discard this PR and open a new one in a new branch?

**Comment by harirakr** [comment]
Merged PR #594

**Pull request closed without merging**